### ISSUE TITLE
Odd highlighting in editor in dart-code-metrics codebase

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#4454]
While investigating #4436, I noticed this odd highlighting:

![Mar-23-2023 10-00-09](https://user-images.githubusercontent.com/1078012/227168671-52e266b1-1eee-4085-b252-33f70a3eb152.gif)

```
==> {
	"jsonrpc": "2.0",
	"id": 213,
	"method": "textDocument/documentHighlight",
	"params": {
		"textDocument": {
			"uri": "file:///Users/danny/Dev/TestStuff/dart-code-metrics/lib/src/analyzers/lint_analyzer/rules/rules_factory.dart"
		},
		"position": {
			"line": 167,
			"character": 7
		}
	},
	"clientRequestTime": 1679565486647
}

<== {
	"id": 213,
	"jsonrpc": "2.0",
	"result": [
		{
			"range": {
				"end": {
					"character": 16,
					"line": 168
				},
				"start": {
					"character": 70,
					"line": 166
				}
			}
		}
	]
}
```

**Comment by DanTup** [comment]
@bwilkerson the current implementation of Occurrences assumes that all occurrences have the same length:

```dart
class Occurrences {
  List<int> offsets;
  int length;
}
```

This doesn't handle constructors referenced using `".new"` (3 vs either 0 or the length of the class name). Some possible fixes:

1. Add a new `List<int> lengths` that has the same length as `offsets` that can be used instead (this is simplest but the API would allow the lengths to be different)
2. Add a `List<Location> location` that contains all ranges (this bloats the payload a little, since `Location` has `file`/`startLine`/`startColumn`/`endLine`/`endColumn`, although having line/col isn't necessarily a bad thing)
3. As above, but create a new class `Occurrence` (or `Range` or similar) that is just `offset`/`length` instead of using `Location`

Another option (if we wanted to avoid changing the protocol and updating IntelliJ/AS) would be to drop "new" from the results and only support it for LSP (although we'd need some tweaks along the lines of the above to return the lengths to use in the LSP handler).

**Comment by bwilkerson** [comment]
Given the following code:
```
class C {
  C();
  C.named();
}

class D extends C {
  D() : super();
}

void f() {
  [
    C(),
    C.new(),
    C.named(),
  ];
}
```

Server is currently highlighting the 'C' and not the '.new'. It's also highlighting the 'C' and not the '.named' whenever some reference to 'C' is selected. Effectively, it's treating all of the 'C's as references to the class, not references to the constructor.

That doesn't seem ideal, but it also doesn't seem terrible. Definitely something we should look into at some point, but I'm not sure it has a very high priority.

If we were going to solve it, then my inclination (as a general principle) would be to replace the uses of the legacy protocol object with a protocol-agnostic object that has all of the information needed by both protocols.

**Comment by DanTup** [comment]
@bwilkerson using a single-character class name slightly masks the issue here. If you make the class name much longer, when your cursor is in `new` you get this (which initially made me think server had gotten out of sync):

<img width="295" alt="Screenshot 2023-03-27 at 16 44 08" src="https://user-images.githubusercontent.com/1078012/227993497-cef7474b-6da0-4746-acbe-997a2734479e.png">

The highlighting includes `new(),\n    C` (which is the length of the class name).

As a short-term fix, we could just exclude `new` constructor references from the occurrences (this would avoid any highlighting in this case), and then as a proper fix later do as you suggest and introduce a new type (though we'd still need to decide how to handle the original protocol - either protocol changes or just always dropping `new` there?).

**Comment by bwilkerson** [comment]
Oh, I see. Definitely worse than I'd realized.

I'm not sure whether the loss of functionality is better or worse than the highlighting issue. I'm kind of inclined at this point to just leave the highlighting issue as-is until we can do a proper fix, which probably won't take that long to implement.

As for the legacy protocol, I'd want to get Alex's input because I don't know what IntelliJ supports on their side.

**Comment by DanTup** [comment]
You're right, it's probably better to just fix entirely. Although whatever we do, we'd need to decide whether to have the highlighting issue or drop the `new` references for any clients of the legacy protocol that haven't been updated to use any new `length` support we add for occurrences.

@alexander-doroshko if the legacy protocol is updated so that we can provide occurrences with different lengths (eg. if the cursor is in `new` in `FooClass.new`, it could highlight the constructor `FooClass();` as a reference to match `new`), or do all occurrences have to be the same length/text as the DAS protocol currently assumes? (I'm assuming IntelliJ is using Occurrences from the server).

**Comment by alexander-doroshko** [comment]
IntelliJ's Dart plugin doesn't use `Occurrences` from the server. Occurrences highlighting works using IntelliJ's internal (language-agnostic) mechanism, but based on the references (aka navigation) info from the server. So it works only when the reference and the target have the same text.

I think we want to switch to the `Occurrences` from the server eventually.

**Comment by alexander-doroshko** [comment]
> If you did, would it support different lengths?

Yes if we have lengths in `Occurrences`.

**Comment by DanTup** [comment]
I started implementing this by adding additional data for lengths, but discovered that for the original protocol things are a little different to LSP and it's not clear to me whether constructors and type names should be bundled together.

For example, in this code:

```dart
class MyClass {                  // 1
  MyClass();                     // 2
  MyClass a = MyClass.new();     // 3
  MyClass b = new MyClass();     // 4
  var c = MyClass;               // 5
}
```

Should `MyClass` on line 1 and 2 be considered occurrences of the same thing (the name of this class) or different (a class and a constructor)?

If we treat them all the same, it might seem odd for ".new" to be highlighted as an occurrence of `class MyClass`. If we treat them separately, that's different to what we do today and might be confusing.

I'm wondering if we should just exclude ".new" entirely, and stick to occurrences always being names? This also avoids some complication that analyzer plugins might be providing occurrences (without lengths) that we need to merge with servers (which means we trade them up using the elements length, and then convert them all back down so offsets are provided to clients for compatibility).

The only niggle with this is that if your cursor is in ".new" you don't see occurrences highlighted - but since that's broken today and you can easily move to the type name, I'm not sure if that's a big problem.

Any thoughts?

**Comment by gnprice** [comment]
> Should `MyClass` on line 1 and 2 be considered occurrences of the same thing (the name of this class) or different (a class and a constructor)?

Treating them separately seems appropriate to me. In my experience it's very common that I want to find specifically the spots where a constructor is called, which may be just one or two sites while the class name is referenced in dozens of places as the type of a field or a parameter or so on. (For example that combination seems to be very common in the Flutter framework.) So it's quite valuable to be able to look just at the occurrences of the constructor.

Treating them separately seems to be the existing behavior of the IntelliJ Dart plugin's occurrence highlighting, and I regularly find that useful for the same reason.

Conversely when one wants all the references to the class name in general, the `class MyClass` is usually just a few lines away.

**Comment by DanTup** [comment]
I had another quick look at this today and it seems the original bug is no longer there, because `new` is always highlighted as if it's different to the constructor (and constructors and classes are the same):

## Cursor on "new"

![Image](https://github.com/user-attachments/assets/a0aa194e-edf3-4679-b0f4-99aa38d848bb)

## Cursor on "AAAAAAAAAAAAAAAAAAAAAAA" default constructor

![Image](https://github.com/user-attachments/assets/e346dfec-7cf3-456b-9733-d1bff40fe8e7)

Before changing anything, I'd like to check we agree on which symbols are grouped together. My suggestion would be as below - I don't think there's anything surprising but the type name might sometimes by "the type name" and sometimes be "the constructor". I think generally the colours here match what Go-to-Definition would do, with one exception - Go-to-Definition on a default _implicit_ constructor would go to the class - however I don't think we should group them together for highlights.

![Image](https://github.com/user-attachments/assets/2446d5a6-1692-4aab-894f-84f859baae79)

Orange boxes: The type name
Lime Green boxes: The default constructor
Blue boxes: The constructor named "named"

@bwilkerson I'm interested in your opinion on this. The ones I think are less obvious are 2 (it feels similar to 3a in some ways, but we treat it differently here) and 6b (similarly).

**Comment by bwilkerson** [comment]
I agree with the choices you made. I think there are clear guidelines that make them more obvious choices, and I'll try to write something down to explain the guidelines.

**Comment by DanTup** [comment]
Thanks - I think I have it working as described above now, working on a CL at https://dart-review.googlesource.com/c/sdk/+/406822.

https://github.com/user-attachments/assets/ca01d733-347a-4796-8ddc-28136388c3bb

For the legacy protocol we can't support occurrences with multiple lengths (eg. `A()` and `A.new()`) in a backwards-compatible way so I'm filtering out occurrences with a different length there. Since IntelliJ is currently using its own implementation I don't think it makes a difference there, but we might want to consider updating the protocol to support that (or, having IntelliJ use LSP-over-Legacy if/when it moves to server-provided occurrences).

**Comment by DanTup** [comment]
> If IntelliJ isn't using the legacy protocol, then we should remove support for the protocol in order to reduce maintenance costs.

Makes sense. I've opened https://github.com/dart-lang/sdk/issues/60043 to track this and listed this and Signature Help as possible candidates (there are likely more, so determining what's unused is possibly the first step - I don't know if there analytics that might help with that?).

As for this issue, this is now resolved via https://github.com/dart-lang/sdk/commit/4be918837925009116d1c0cecac8a6d9f5ae4fa7.
